packages/firestore/src/model/mutation.ts

```diff
@@ -239,89 +252,187 @@ export abstract class Mutation {
   abstract readonly type: MutationType;
   abstract readonly key: DocumentKey;
   abstract readonly precondition: Precondition;
+}
 
-  /**
-   * Applies this mutation to the given MaybeDocument or null for the purposes
-   * of computing a new remote document. If the input document doesn't match the
-   * expected state (e.g. it is null or outdated), an `UnknownDocument` can be
-   * returned.
-   *
-   * @param maybeDoc The document to mutate. The input document can be null if
-   *     the client has no knowledge of the pre-mutation state of the document.
-   * @param mutationResult The result of applying the mutation from the backend.
-   * @return The mutated document. The returned document may be an
-   *     UnknownDocument if the mutation could not be applied to the locally
-   *     cached base document.
-   */
-  abstract applyToRemoteDocument(
-    maybeDoc: MaybeDocument | null,
-    mutationResult: MutationResult
-  ): MaybeDocument;
+/**
+ * Applies this mutation to the given MaybeDocument or null for the purposes
+ * of computing a new remote document. If the input document doesn't match the
+ * expected state (e.g. it is null or outdated), an `UnknownDocument` can be
+ * returned.
+ *
+ * @param mutation The mutation to apply.
+ * @param maybeDoc The document to mutate. The input document can be null if
+ *     the client has no knowledge of the pre-mutation state of the document.
+ * @param mutationResult The result of applying the mutation from the backend.
+ * @return The mutated document. The returned document may be an
+ *     UnknownDocument if the mutation could not be applied to the locally
+ *     cached base document.
+ */
+export function applyMutationToRemoteDocument(
```

Comment:
Apologies in advance for this drive by comment :-).

Why is this pattern better than just having the lite SDK create the `Write` protos directly? This seems like way more code added and in the end the lite SDK still ends up with an intermediate object that contributes to the heft of the SDK.

When simplifying documents in the other SDKs we made them just wrap the protos. It seems like we could take a similar approach here: have the lite SDK to just build the protos and pass them down. The heavier-weight SDKs could reuse that code but then wrap the protos in classes that add local behavior.

This way we'd still end up with code sharing between the SDKs, the lite SDK ends up optimally lite (with every API call is just translation of arguments into protos and a REST invocation), but we don't end up having to deal with these instanceof ladders for behavior that doesn't fit in the lite SDK.

Note: this comment is really a question. I don't really understand why you disliked the outcome of #3327 and I don't understand what value mutations have in the lite SDK. If you're convinced this path is the way forward I'll be happy with that outcome but I just want to make sure because this pattern is kinda gross.

Reply:
The reason I don't like #3327:
- It duplicates the Proto encoding for users that use the Write-proto based Transactions and non-transactional writes. This leads to a general size increase for all the main SDK and firestore-exp.
- It makes it impossible to share the WriteBatch between the firestore-exp and the Lite SDK.
- It adds a lot of functions with names that I am not happy with.